Index: rpm-4.9.1/fileattrs/perl.attr
===================================================================
--- rpm-4.9.1.orig/fileattrs/perl.attr
+++ rpm-4.9.1/fileattrs/perl.attr
@@ -1,3 +1,3 @@
-%__perl_requires	%{_rpmconfigdir}/perl.req
+#%__perl_requires	%{_rpmconfigdir}/perl.req
 %__perl_magic		^.*perl .*$
 %__perl_flags		exeonly
Index: rpm-4.9.1/fileattrs/perllib.attr
===================================================================
--- rpm-4.9.1.orig/fileattrs/perllib.attr
+++ rpm-4.9.1/fileattrs/perllib.attr
@@ -1,3 +1,3 @@
 %__perllib_provides	%{_rpmconfigdir}/perl.prov
-%__perllib_requires	%{_rpmconfigdir}/perl.req
+#%__perllib_requires	%{_rpmconfigdir}/perl.req
 %__perllib_magic	^Perl[[:digit:]] module source.*
