--- ./autodeps/linux.prov.orig	2011-05-11 15:59:31.000000000 +0000
+++ ./autodeps/linux.prov	2011-05-11 16:39:09.000000000 +0000
@@ -11,7 +11,7 @@ filelist=($(printf "%s\n" "${filelist[@]
 solist=($(printf "%s\n" "${filelist[@]}" | grep "\\.so" | grep -v "^/lib/ld.so" | \
 	  tr '\n' '\0' | xargs -0 -r file -L | grep "ELF.*shared object" | \
 	  cut -d: -f1))
-pythonlist=
+pythonlist=($(printf "%s\n" "${filelist[@]}" | egrep '/usr/bin/python.\..$'))
 tcllist=
 monolist=($(printf "%s\n" "${filelist[@]}" | egrep "\\.(exe|dll)\$"))
 mimetypelist=($(printf "%s\n" "${filelist[@]}" | egrep "\\.(desktop)\$"))
@@ -55,8 +55,8 @@ done | sort -u
 
 #
 # --- Python modules.
-[ -x /usr/lib/rpm/python.prov -a -n "$pythonlist" ] &&
-    printf "%s\n" "${pythonlist[@]}" | /usr/lib/rpm/python.prov | sort -u
+[ -x /usr/lib/rpm/pythondeps.sh -a -n "$pythonlist" ] &&
+    printf "%s\n" "${pythonlist[@]}" | /usr/lib/rpm/pythondeps.sh -P | sort -u
 
 #
 # --- Tcl modules.
--- ./autodeps/linux.req.orig	2011-05-11 14:28:45.000000000 +0000
+++ ./autodeps/linux.req	2011-05-11 16:39:09.000000000 +0000
@@ -31,7 +31,7 @@ liblist=($(printf "%s\0" "${filelist[@]}
 
 interplist=()
 perllist=()
-pythonlist=()
+pythonlist=($(printf "%s\n" "${filelist[@]}" | egrep '/usr/lib[^/]*/python.\..'))
 tcllist=()
 monolist=($(printf "%s\n" "${filelist[@]}" | egrep "\\.(exe|dll)(\\.config)?\$"))
 
@@ -127,8 +127,8 @@ done | sort -u
 
 #
 # --- Python modules.
-[ -x /usr/lib/rpm/python.req -a -n "$pythonlist" ] && \
-    printf "%s\n" "${pythonlist[@]}" | /usr/lib/rpm/python.req | sort -u
+[ -x /usr/lib/rpm/pythondeps.sh -a -n "$pythonlist" ] && \
+    printf "%s\n" "${pythonlist[@]}" | /usr/lib/rpm/pythondeps.sh -R | sort -u
 
 #
 # --- Tcl modules.
